### PR TITLE
Haskell/norg c++14 compilation

### PR DESCRIPTION
**Pull request by theHamsta**
I don't know yet whether this is a good idea or whether the parser themselves should just have their own `Makefile` or `npm` task but we could at least increase the required C++ standard by having there compile commands.

A better solution would of course be to have a unified compilation, release and installation mechanism in upstream tree-sitter

https://github.com/nvim-treesitter/nvim-treesitter/pull/1766 @JoeyGrajciar

### CODE REVIEW

**Comment by theHamsta** [comment]
About the failures on Windows, don't you have to manually export symbols to make them accessible on Windows using MSVC. How did this work until now? Why is it failing now on GCC and MSVC?

File: scripts/compile_parsers.makefile
Comment (theHamsta) [inline scripts/compile_parsers.makefile:13]: for our current parsers, c++14 would be enough

File: scripts/compile_parsers.makefile
Comment (theHamsta) [inline scripts/compile_parsers.makefile:16]: Does stdc++ exist on Mac? Or will clang just do some compat magic?

Comment (JoeyGrajciar) [reply scripts/compile_parsers.makefile:18]: Tried to use `-lstdc++` on my M1 and it works with Apple clang 12.0.5

Comment (theHamsta) [reply scripts/compile_parsers.makefile:18]: Yes, clang just tries to behave like GCC

**Comment by theHamsta** [comment]
@JoeyGrajciar any idea why this fails for windows gcc/clang https://github.com/nvim-treesitter/nvim-treesitter/runs/4373980291?check_suite_focus=true ? though it says that it has installed norg.

**Comment by JoeyGrajciar** [comment]
@theHamsta yes it is weird but error is correct `norg.so` file is missing in the artifacts.

**Review by mrossinek — COMMENTED** [review]
Hi there! Thanks for including `norg` in this PR. We moved to an organization some time ago so I'm suggesting we update these links here. I'm one of the maintainers over there who is rather involved with the TS parser too at this point so I took the liberty of adding myself to the list, too.

File: lua/nvim-treesitter/shell_command_selectors.lua
Comment (JoeyGrajciar) [inline lua/nvim-treesitter/shell_command_selectors.lua:97]: @theHamsta This is why it is failing :) `cc` is `clang` so it will go to `make` branch it does not find it, but probably the sequence with `pushd` and `popd` returns 0

Comment (JoeyGrajciar) [reply lua/nvim-treesitter/shell_command_selectors.lua:97]:
```suggestion
  if string.match(cc, "cl$") or string.match(cc, "cl.exe$") or not repo.use_makefile or fn.has "win32" == 1 then
```

**Comment by theHamsta** [comment]
Thanks for helping out!

File: .github/workflows/check-query-files-and-compilation.yml
Comment (JoeyGrajciar) [inline .github/workflows/check-query-files-and-compilation.yml:74]: is this needed?

Comment (theHamsta) [reply .github/workflows/check-query-files-and-compilation.yml]: https://github.com/nvim-treesitter/nvim-treesitter/runs/4386295723?check_suite_focus=true

Comment (JoeyGrajciar) [reply .github/workflows/check-query-files-and-compilation.yml]: okay but that failure is different and not only `tree-sitter-norg`
```
Downloading...

Checking out locked revision

The system cannot find the path specified.

fatal: reference is not a tree: 74a721e8eed1a4a25cf495d45974ba24f315f81a

Error detected while processing command line:

Error while checking out revision

Failed to execute the following command:

{

  cmd = "git",

  err = "Error while checking out revision",

  info = "Checking out locked revision",

  opts = {

    args = { "checkout", "74a721e8eed1a4a25cf495d45974ba24f315f81a" },

    cwd = "C:\\Users\\runneradmin\\AppData\\Local\\nvim-data\\tree-sitter-prisma"

  }

}

Downloading...

Checking out locked revision

The system cannot find the path specified.

error: pathspec 'main' did not match any file(s) known to git

Error while checking out revision

Failed to execute the following command:

{

  cmd = "git",

  err = "Error while checking out revision",

  info = "Checking out locked revision",

  opts = {

    args = { "checkout", "main" },

    cwd = "C:\\Users\\runneradmin\\AppData\\Local\\nvim-data\\tree-sitter-norg"

  }

}

Downloading...

Checking out locked revision

The system cannot find the path specified.

fatal: reference is not a tree: 111b07762e86efab9a918b7c721f720c37e76b0a

Error while checking out revision

Failed to execute the following command:

{

  cmd = "git",

  err = "Error while checking out revision",

  info = "Checking out locked revision",

  opts = {

    args = { "checkout", "111b07762e86efab9a918b7c721f720c37e76b0a" },

    cwd = "C:\\Users\\runneradmin\\AppData\\Local\\nvim-data\\tree-sitter-typescript"

  }

}
```
and I think also 2-3 other failed with same reason that there was some issue with downloading

Comment (JoeyGrajciar) [reply .github/workflows/check-query-files-and-compilation.yml]: Also the last run for gcc it was `graphql` and `html` having issues like this

Comment (theHamsta) [reply .github/workflows/check-query-files-and-compilation.yml]: Master has same problem, maybe we should just fetch using curl on CI

**Comment by theHamsta** [comment]
@stsewd I think this is ready to go. This allows compilation of C++14 parsers like Haskell and Norg using old compilers and the default clang on MacOS 10. The CI failures also happen on other branches

**Comment by theHamsta** [comment]
Let's try curl.

**Comment by theHamsta** [comment]
Maybe this trick would also work: https://github.com/emacs-tree-sitter/elisp-tree-sitter/issues/166#issuecomment-986262035 But makefiles would be a bit more flexible

**Comment by mrossinek** [comment]
Just a quick note from my side: I tested this branch out and everything seems to work just as expected. This is on Linux so maybe we can get someone on Windows and Mac to try as well?

**Comment by theHamsta** [comment]
@mrossinek This does not change the Windows behavior. Compilation on Windows was working before. On Mac, I got positive feedback that this would work. @kyazdani42 can we merge this? This would be maybe an alternative to the makefiles if this works: https://github.com/nvim-treesitter/nvim-treesitter/pull/2094

File: lua/nvim-treesitter/shell_command_selectors.lua
Comment (vigoux) [inline lua/nvim-treesitter/shell_command_selectors.lua:103]: I feel like we should also use the `cxx_standard` key in the `select_compiler_args` for completeness too, do you think this is wise ?

Comment (theHamsta) [reply lua/nvim-treesitter/shell_command_selectors.lua:103]: If it works without C++14, I would leave it that way. I think only Haskell and Norg require C++14 at the moment which works with a recent Windows compiler or Makefiles on Linux/MacOS

Comment (theHamsta) [reply lua/nvim-treesitter/shell_command_selectors.lua:103]: Ah, now I remember. It is not possible when invoking a C compiler with a single invocation. This is why I was using the Makefiles. In theory, it should work when you scope the cxx_standard with `-Xc++` flag. But it didn't work when I tried it.

**Comment by danymat** [comment]
Hello, what's the state of this PR ? Is it ready to merge ?

**Comment by theHamsta** [comment]
I'd say it's ready.

**Comment by theHamsta** [comment]
@stsewd @vigoux is there anything left to do here? I'd say we can merge.

**Comment by theHamsta** [comment]
Haskell is no longer affected.  norg and foam parser would like to have C++14

**Comment by FoamScience** [comment]
I'll revert back to C too (for foam parser) as C++14 portability issues are chasing me around in many places :)

**Comment by theHamsta** [comment]
@clason can you check whether there's something still blocking this PR?

**Comment by clason** [comment]
You rebased this PR on master?

**Comment by theHamsta** [comment]
yes

**Review by clason — APPROVED** [review]
Hmm, I haven't followed the discussion closely, but I trust you (and CI). Let's see if cleanup is necessary ;)

**Pull request merged**
